src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+// InternalCommand, it conceivably will when appropriate
+
+namespace System.Management.Automation
+{
+    #region Keyword Command Class
+
+    /// <summary>
+    /// An abstract class to extend in order to write a keyword specification. Give this class the KeywordAttribute
+    /// in order to use it as a PowerShell DynamicKeyword specification. Any properties with
+    /// the KeywordParameterAttribute will be parameters, while properties with the KeywordPropertyAttribute will
+    /// be keyword properties.
+    /// </summary>
+    public abstract class Keyword
+    {
+        /// <summary>
+        /// Constructs a keyword with null runtime delegates. This is intended to be overridden.
+        /// </summary>
+        protected Keyword()
+        {
+        }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword data object before the AST node
+        /// containing that keyword is parsed
+        /// </summary>
+        public Func<DynamicKeyword, ParseError[]> PreParse { get; set; }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword statement AST immediately after that AST node
+        /// has been parsed
+        /// </summary>
+        public Func<DynamicKeywordStatementAst, ParseError[]> PostParse { get; set; }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword statement AST node at semantic check time (to perform
+        /// any user-specified semantic checks)
+        /// </summary>
+        public Func<DynamicKeywordStatementAst, ParseError[]> SemanticCheck { get; set; }
+    }
+
+    #endregion /* Keyword Command Class */
+
+    #region Dynamic Keyword Specification Attributes
+
+    /// <summary>
+    /// Specifies that a class describes a DynamicKeyword for PowerShell
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public class KeywordAttribute : ParsingBaseAttribute
+    {
+        // Default attribute parameter values
+
+        /// <summary>
+        /// Default keyword name mode
+        /// </summary>
+        public const DynamicKeywordNameMode DefaultNameMode = DynamicKeywordNameMode.NoName;
+
+        /// <summary>
+        /// Default keyword body mode
+        /// </summary>
+        public const DynamicKeywordBodyMode DefaultBodyMode = DynamicKeywordBodyMode.Command;
+        
+        /// <summary>
+        /// Default keyword use mode
+        /// </summary>
+        public const DynamicKeywordUseMode DefaultUseMode = DynamicKeywordUseMode.OptionalMany;
+
+        /// <summary>
+        /// Default keyword resource name -- TODO: This might be better as null
+        /// </summary>
+        public static readonly string DefaultResourceName = null;
+
+        /// <summary>
+        /// Keyword defaults to being a marshalled call
+        /// </summary>
+        public const bool DefaultIsDirectCall = false;
+
+        /// <summary>
+        /// Keyword defaults to being part of the AST
```

Comment:
Yes, quite right -- it's in the compiler that a `MetaStatement` keyword generates an empty expression, and I think I confused myself.